fourshores-company/greenpot-backend
# User Role Setting

## PR Overview
This PR adds a User Role setting feature to the greenpot application. The user role setting allows for permissions to be granted to individual users of the application. 

Roles
- Admin
- user

Also added
-  added admin validations
- added verify roles middleware
- created route for role setting
- created user controller for updating roles
- edited swagger docs
- user tests

### Note
The verify roles middleware takes in an array of numbers as `roleIds`; these represent the permissions allowed on the particular route. this dynamically allows any configuration of permissions needed per route, and excludes any permissions not explicitly stated.

Permission combos can be added to object in `src/utils/permissionData`

## Testing
- Clone the repo and cd into the project folder
- Update or copy `.env` file from `.env.sample`
- Checkout to `ft-user-search` branch and run `npm install`
- - Run `npm run migrate:reset` && `npm run migrate` to create tables in your 
 database and seed the database tables with the necessary app configs.
- Run `docker-compose up --build` to build the docker image and spin up all the necessary services
- As an alternative to running the app on docker, you can run `npm run start:dev`
- Launch a *PATCH* call for`http://localhost:3000/v1.0/api/user/role`on Postman, make sure a user is already logged in with his/her token in the cookies.

- input object
```
{
	"roleId": 2,
	"adminKey": "[ADMIN_KEY]"
}
```
### Note for Testing
make sure the `ADMIN_KEY` matches that set in your .env file


## Screenshots
<img width="459" alt="Screenshot 2019-10-18 at 08 06 04" src="https://user-images.githubusercontent.com/37340699/67080717-b328f780-f18d-11e9-8df6-283f06c60758.png">
<img width="459" alt="Screenshot 2019-10-18 at 08 05 27" src="https://user-images.githubusercontent.com/37340699/67080727-b91ed880-f18d-11e9-9ea0-99817ea03e75.png">
<img width="811" alt="Screenshot 2019-10-18 at 07 58 08" src="https://user-images.githubusercontent.com/37340699/67080738-c045e680-f18d-11e9-9e4c-227c7d87f1c5.png">
<img width="783" alt="Screenshot 2019-10-18 at 07 53 34" src="https://user-images.githubusercontent.com/37340699/67080751-c63bc780-f18d-11e9-877f-964a16f2e0eb.png">
